helix-core/src/auto_pairs.rs

```diff
@@ -100,64 +115,337 @@ fn handle_open(
 
                 offs += 2;
 
-                (pos, pos, Some(pair))
+                (start_head, start_head, Some(pair))
             }
         }
     });
 
-    transaction.with_selection(Selection::new(ranges, selection.primary_index()))
+    let t = transaction.with_selection(Selection::new(end_ranges, selection.primary_index()));
+    debug!("auto pair transaction: {:#?}", t);
+    t
 }
 
 fn handle_close(doc: &Rope, selection: &Selection, _open: char, close: char) -> Transaction {
-    let mut ranges = SmallVec::with_capacity(selection.len());
+    let mut end_ranges = SmallVec::with_capacity(selection.len());
 
     let mut offs = 0;
 
-    let transaction = Transaction::change_by_selection(doc, selection, |range| {
-        let pos = range.head;
-        let next = next_char(doc, pos);
+    let transaction = Transaction::change_by_selection(doc, selection, |start_range| {
+        let start_head = start_range.head;
+        let next = next_char(doc, start_head);
+        let end_head = start_head + offs + close.len_utf8();
 
-        let head = pos + offs + close.len_utf8();
-        // if selection, retain anchor, if cursor, move over
-        ranges.push(Range::new(
-            if range.is_empty() {
-                head
-            } else {
-                range.anchor + offs
-            },
-            head,
-        ));
+        let end_anchor = if start_range.is_empty() {
+            end_head
+        } else {
+            start_range.anchor + offs
+        };
+
+        end_ranges.push(Range::new(end_anchor, end_head));
 
         if next == Some(close) {
-            //  return transaction that moves past close
-            (pos, pos, None) // no-op
+            // return transaction that moves past close
+            (start_head, start_head, None) // no-op
         } else {
             offs += close.len_utf8();
 
             // TODO: else return (use default handler that inserts close)
-            (pos, pos, Some(Tendril::from_char(close)))
+            (start_head, start_head, Some(Tendril::from_char(close)))
         }
     });
 
-    transaction.with_selection(Selection::new(ranges, selection.primary_index()))
+    transaction.with_selection(Selection::new(end_ranges, selection.primary_index()))
 }
 
-// handle cases where open and close is the same, or in triples ("""docstring""")
-fn handle_same(_doc: &Rope, _selection: &Selection, _token: char) -> Option<Transaction> {
-    // if not cursor but selection, wrap
-    // let next = next char
-
-    // if next == bracket {
-    //   // if start of syntax node, insert token twice (new pair because node is complete)
-    //   // elseif colsedBracketAt
-    //      // is_triple == allow triple && next 3 is equal
-    //      // cursor jump over
-    // }
-    //} else if allow_triple && followed by triple {
```

Comment:
The proposed algorithm here from codemirror also handled triples correctly (for docstrings for example), can you implement it/add a TODO?

Reply:
The CodeMirror code for this looks pretty hand tailored specifically for triplets, and I'm planning to cover this when I change the open/close pairs to be arbitrary strings, so it can place the cursor in the middle for n-char pairs. There's a TODO for it near the top.

Reply:
It deals with triplets specially so that it can handle `""|`  -> `"""|"""`, so that it understands the " is a part of a triplet and not just closing a pair.

Reply:
Yep. And when pairs are arbitrary strings, it will have a similar understanding. It will additionally be able to handle `{|}` → `{%|%}`, `f'|` → `f'|'`, `ba|` → `bar|quux` or anything else. Handling triplets specially would be redundant.